python/hail/tests/test_typecheck.py

```diff
@@ -259,4 +259,24 @@ def foo(f):
 
         self.assertRaises(TypeError, lambda: foo(l1))
         self.assertRaises(TypeError, lambda: foo(l2))
-        foo(l3)
\ No newline at end of file
+        foo(l3)
+
+    def test_complex_signature(self):
+
+        @typecheck(a=int, b=int, c=int, d=int, e=int)
```

Comment:
how does this test/typecheck work? (and maybe the test could have these things be different types to make it clearer?)

Reply:
The switch to using functools instead of decorator made things a bit more complicated, because the wrapped functions are actually *args, **kwargs instead of being code-generated as the same signature. All these invocations of `f` pass arguments different ways: as args, as kwargs, as default arguments. I wanted to make sure that all the various ways of calling this function succeed.

Reply:
it'd be nice to have different arg types to makes sure that the decorator is checking the right argument in the right place.